Python 3.12 support

laclouis5 opened the pull request:
This PR adds support for Python 3.12 and drop support for Python 3.8. I updated the minimal dependencies in the `pyproject.toml` to match the minimal dependencies specified by Scipy, Matplotlib and Numpy for Python >= 3.9.

I updated the top-level `__init__.py` to remove the now obsolete use of the `importlib` package.

I also removed the `include = ["tests"]` line in the `pyproject.toml` file since I wasn't sure why the `test` package would be included in the wheel (this should be the case). 

The `.gitignore` has also been improved (ignore macOS related files, etc.).

I did not increases the package version number yet but I updated the project metadata (Python compatibility, etc.).

ajhynes7 commented on .gitignore:
```diff
@@ -1,16 +1,178 @@
+*.code-workspace
+
+# Local History for Visual Studio Code
+.history/
```
It looks like you replaced this repo's `.gitignore` with your own. Can you undo this change please?

laclouis5 replied:
I reverted the changes to keep the PR focused but this was intentional. This was a `.gitignore` created with the tool gitignore.io with the keywords "Python", "VisualStudioCode" and "macOS" and which includes lots of artefacts that may exist in many configurations, which is quite useful for such colarboration projects.

ajhynes7 commented on .github/workflows/main.yml:
```diff
@@ -9,7 +9,7 @@ jobs:
     strategy:
       matrix:
-        python-version: [3.8, 3.9, "3.10"]
+        python-version: ["3.9", "3.10", "3.12", "3.12"]
```
You have `"3.12"` twice here.

ajhynes7 reviewed (changes requested):
Hi @laclouis5,

I'm fine with adding support for Python 3.12, but the end of life for Python 3.8 isn't until [Oct. 2024](https://devguide.python.org/versions/#supported-versions), so I'd like to keep supporting 3.8 in this package until then.

CristianoPizzamiglio commented:
> Hi @laclouis5,
> 
> I'm fine with adding support for Python 3.12, but the end of life for Python 3.8 isn't until [Oct. 2024](https://devguide.python.org/versions/#supported-versions), so I'd like to keep supporting 3.8 in this package until then.

I totally agree with @ajhynes7, I really would like have Python 3.8 supported until EOL. Thanks!

laclouis5 commented:
> > Hi @laclouis5,
> > I'm fine with adding support for Python 3.12, but the end of life for Python 3.8 isn't until [Oct. 2024](https://devguide.python.org/versions/#supported-versions), so I'd like to keep supporting 3.8 in this package until then.
> 
> I totally agree with @ajhynes7, I really would like have Python 3.8 supported until EOL. Thanks!

Yes this makes sense, I added back Python 3.8 support!

ajhynes7 commented:
The `[gh-actions]` section of `tox.ini` needs to be updated as well.

laclouis5 commented:
> The `[gh-actions]` section of `tox.ini` needs to be updated as well.

I updated it, though that wasn't the cause of the GitHub Actions failure since this was an issue with `distutils` being removed in Python 3.12 (I added a `setuptools` dependency in the doc requirements to solve the issue).

I don't think that the `[gh-actions]` is used properly by GitHub actions because it was outdated: it still tested for Python 3.7 but there was no run for Python 3.7 in the Actions summary. The Python versions used for the GitHub Actions seems to be the ones in the `.github/workflows/main.yml`.

ajhynes7 commented:
>I don't think that the [gh-actions] is used properly by GitHub actions because it was outdated

Yeah it's been a while since I set that up. I should take a look sometime to see if that section is still needed.

ajhynes7 commented:
Thanks @laclouis5! Are you planning on doing another PR to increment the version, or do you want me to do that?

laclouis5 commented:
@ajhynes7 Yes I can do it. Would incrementing the patch version be fine (i.e 7.0.1) since this PR does not change the API or affect backward compatibility?

ajhynes7 commented:
@laclouis5 Ok thanks! It would be `7.1.0` because it's a feature, not a bug fix.
